File: 19-RemoveNthNodeFromEndOfList/19-RemoveNthNodeFromEndOfList.py
# Last updated: 12/6/2025, 5:55:06 am
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if head.next == None: return None
        cnt = 0
        def rec(node):
            nonlocal cnt
            if node == None: return 0
            index = rec(node.next)
            cnt += 1
            if index == n:
                node.next = node.next.next
                return 31
            return index+1
        index = rec(head)
        if index == cnt: return head.next
        return head

File: 19-RemoveNthNodeFromEndOfList/19-RemoveNthNodeFromEndOfList.cpp
// Last updated: 12/6/2025, 5:55:12 am
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *dummy = new ListNode(0);
        ListNode *node = head;
        int cnt = 0;
        while (true){
            if (node->next == NULL) {cnt++; break;}
            cnt++;
            node = node->next;
        }int valid = cnt - n;
        node = head;
        ListNode *ret = dummy;
        for (unsigned int i=0; i<cnt; i++){
            if (i != valid){
                ret->next = new ListNode(node->val);
                ret = ret->next;
                node = node->next;
            }else node = node->next;
        }ret = dummy->next;
        delete dummy, node;
        return ret; 
    }
};